### DISCUSSION TITLE
Plotting with Geopandas

### DISCUSSION THREAD

**Discussion by travisz09** [ua-gist604b-f24/syllabus#13] [General]
Hi all,

I am working through assignment 4-1-Geopandas-Intro. Everything is going well, and I feel I have a good understanding of Geopandas, but I did have one question about plotting.

When it comes to plotting multiple layers on a single map, it makes sense to pass one 'layer' as an axis argument for a second layer, thereby stacking your layers to make a single map. But then in the tutorial I saw this syntax.

> \# saving the first plot as an axis and setting alpha (transparency) to 0.5
> ax = gdf["buffered"].plot(alpha=0.5)
> \# passing the first plot as an axis to the second
> gdf["buffered_centroid"].plot(ax=ax, color="red", alpha=0.5)
> \# passing the first plot and setting linewidth to 0.5
> gdf["boundary"].plot(ax=ax, color="white", linewidth=0.5)

At first, I thought this might be a typo, as I would expect that the syntax to be:

> ax = gdf["buffered"].plot(alpha=0.5)
> ax = gdf["buffered_centroid"].plot(ax=ax, color="red", alpha=0.5)
> gdf["boundary"].plot(ax=ax, color="white", linewidth=0.5)

Where the first layer is assigned to ax, which is passed to the second layer, overwriting ax, then both those layers are passed to the third layer. To me, the expected output of the first example would be two separate plots, with the second plot only showing the 'boundary' layer. However, I was surprised to see that the first example worked correctly, outputting a single plot with three layers (b.t.w. my example also works). This behavior was not explained well in the tutorial, and I have not seen anything that adequately explains how this works.

By comparison:

> gdf["buffered"].plot(alpha=0.5)
> gdf["buffered_centroid"].plot(color="red", alpha=0.5)
> gdf["boundary"].plot(color="white", linewidth=0.5)

does output 3 separate plots.

Does anybody have any insight into this? I am happy to accept that this is just the way it is, but can anybody explain it simply? Is anybody else confused by this syntax, or is it just me?

Thanks all, for reading,
Travis

**Comment by aaryno** [comment]
Travis,

If I understand your question correctly, you are asking about the difference between these three cases (simplified to highlight the important elements:

**A)**
```
gdf("buffered").plot()
gdf("buffered_centroid").plot()
```
**B)**
```
ax = gdf("buffered").plot()
gdf("buffered_centroid").plot(ax=ax)
```
**C)**
```
ax = gdf("buffered").plot()
ax = gdf("buffered_centroid").plot(ax=ax)
```

This will make more sense if we look at what `plot()` is actually doing. Take a look at the docs: https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoDataFrame.plot.html. You'll see two important items regarding the axes. First, there is an **argument** for `ax`: 
<img width="405" alt="image" src="https://github.com/user-attachments/assets/548bb855-f492-4db1-a250-88644fd0ca3b">

This is an _optional_ argument that, if specified, will result in the `plot()` being applied to an existing `ax axes` instance.

Second, there is an instance of `axes` in the **return**:
<img width="340" alt="image" src="https://github.com/user-attachments/assets/35598bf4-ea08-422e-89a9-cabc9ea73369">

So in **A)**, the first call creates an instance of `axes` but we ignore the returned `axes` object. Thus, in the second call, since there is no `ax` argument specified, it created a new `axes` object (and also ignores the return). 

In **B)**, however, we assign the returned `axes` object to a variable named `ax` and then pass that into the second call so that a new `axes` is NOT created. Instead, the plot is applied to the same object as the first call.

This is only slightly different from **C)**, where we are doing the same thing, but we are _re-assigning_ the value of `ax` to the returned `axes` object in the second call. _There is no fundamental difference between **B)** and *C)** the way this is written_. That is, the _returned_ `axes` object in the second call is the same instance as the one we passed in as an argument in `ax=ax`. Thus, at the end of the second call, we still have a single `axes` object referenced by the variable named `ax`.  That said, there is some nuance here that should be aware of. When we re-assign an existing variable like that we could face the possibility that the new variable refers to another instance and that our re-assignment actually clobbers or masks the original variable. That's not the case here but you should be aware that variable reassignment can cause confusion if you aren't paying close attention to what's happening to the original variable.

Some more details about `geopandasdataframe.plot()`: https://geopandas.org/en/stable/docs/user_guide/mapping.html

**Reply by travisz09** [reply]
Thank you for this explanation! 

This makes perfect sense to me when thinking about 2 layers. The plot instance is saved to a variable 'ax', and the second plot function calls that named instance and appends it.

My confusion comes when adding a third layer to the mix. How does the program "know" that 'ax' now has been appended with a second layer without assigning it to a variable. For clarity, lets use unique variables for the axes, and avoid overwriting.

> \# Baselayer
> ax1 = gdf["buffered"].plot(alpha=0.5)
> \# Layer 2
> ax2 = gdf["buffered_centroid"].plot(ax=ax1, color="red", alpha=0.5)
> \# Top Layer
> gdf["boundary"].plot(ax=ax1, color="white", linewidth=0.5)

![image](https://github.com/user-attachments/assets/e070dc4e-5d17-45d5-b279-d42fddb1e115)

Even when explicitly assigning the output of the second plot layer to a new variable 'ax2', 'ax1' is still appended to include both prior layers. Not only do I find this counterintuitive, I could see instances where this would be incredibly annoying. Say that you have a default baselayer which you would like to call repeatedly for a map series. Instead of establishing a persistent variable 'ax' (or perhaps 'base') that you could reuse indefinitely, you would have to reestablish your 'ax' for every map in your series.

By comparison, I find the logic much more intuitive in R using the "Grammar of Graphics" package ggplot2. Using the packages, 'terra', 'tidyterra', and 'ggplot2' I can create a series of maps like so:

> \# Baselayer using bourough buffers data (plot A)
> base = ggplot() +  \# initiate plot instance, assigned to var 'base'
>        &nbsp;&nbsp;geom_spatvector(data = buffer, fill = 'blue', alpha = 0.5) +  \# layer data and aesthetics
>        &nbsp;&nbsp;theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))  \# additional plotting options
> plot(base)  \# show plot
> 
> \# Baselayer with buffered centroids (plot B)
> base + geom_spatvector(data = bufferedCentroids, fill = 'red', alpha = 0.5)
> 
> \# Baselayer with bourough polygons (plot C)
> base + geom_spatvector(data = nybb, fill = NA)
> 
> \# All three layers combined (plot D)
> base +
>        &nbsp;&nbsp;geom_spatvector(data = bufferedCentroids, fill = 'red', alpha = 0.5) +
>        &nbsp;&nbsp;geom_spatvector(data = nybb, fill = NA)

![Rplot02](https://github.com/user-attachments/assets/3dbcc1d8-65a4-4b0e-a856-16508cb2ea25)

I think this is just a fundamentally different way of thinking about the plotting environment, and while I can see why someone might like the geopandas syntax, I don't think I will ever be fully satisfied by that logic. That being said, I can accept that that is the way it is and that it does make some sense, depending on your perspective.

Thank you for your reply, and for providing a space for me to explore these ideas. I hope this does not come off as argumentative, as I am trying to grapple with these concepts and understand which syntax I prefer, and why. I am curious if anyone else agrees or disagrees with my perspective, and I hope that this post has been informative and is not further confusing the topic.

Best,
Travis
